pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go

```diff
@@ -0,0 +1,791 @@
+		claim := state.claims[index]
+		if len(claim.Status.ReservedFor) == 0 ||
+			len(claim.Status.ReservedFor) == 1 && claim.Status.ReservedFor[0].UID == pod.UID {
+			claim := state.claims[index].DeepCopy()
+			claim.Status.DeallocationRequested = true
+			claim.Status.ReservedFor = nil
+			logger.V(5).Info("Requesting deallocation of ResourceClaim", "pod", klog.KObj(pod), "resourceclaim", klog.KObj(claim))
+			if err := state.updateClaimStatus(ctx, pl.clientset, index, claim); err != nil {
+				return nil, statusError(logger, err)
+			}
+			return nil, nil
+		}
+	}
+	return nil, framework.NewStatus(framework.UnschedulableAndUnresolvable, "still not schedulable")
+}
+
+// PreScore is passed a list of all nodes that would fit the pod. Not all
+// claims are necessarily allocated yet, so here we can set the SuitableNodes
+// field for those which are pending.
+func (pl *dynamicResources) PreScore(ctx context.Context, cs *framework.CycleState, pod *v1.Pod, nodes []*v1.Node) *framework.Status {
+	if !pl.enabled {
+		return nil
+	}
+	state, err := getStateData(cs)
+	if err != nil {
+		return statusError(klog.FromContext(ctx), err)
+	}
+	if len(state.claims) == 0 {
+		return nil
+	}
+
+	logger := klog.FromContext(ctx)
+	podScheduling, err := state.initializePodScheduling(ctx, pod, pl.podSchedulingLister)
+	if err != nil {
+		return statusError(logger, err)
+	}
+	pending := false
+	for _, claim := range state.claims {
+		if claim.Status.Allocation == nil {
+			pending = true
+		}
+	}
+	if pending && !haveAllNodes(podScheduling.Spec.PotentialNodes, nodes) {
+		// Remember the potential nodes. The object will get created or
+		// updated in Reserve. This is both an optimization and
+		// covers the case that PreScore doesn't get called when there
+		// is only a single node.
+		logger.V(5).Info("remembering potential nodes", "pod", klog.KObj(pod), "potentialnodes", klog.KObjSlice(nodes))
+		podScheduling = podScheduling.DeepCopy()
+		numNodes := len(nodes)
+		if numNodes > resourcev1alpha1.PodSchedulingNodeListMaxSize {
+			numNodes = resourcev1alpha1.PodSchedulingNodeListMaxSize
+		}
+		podScheduling.Spec.PotentialNodes = make([]string, 0, numNodes)
+		if numNodes == len(nodes) {
+			// Copy all node names.
+			for _, node := range nodes {
+				podScheduling.Spec.PotentialNodes = append(podScheduling.Spec.PotentialNodes, node.Name)
+			}
+		} else {
+			// Select a random subset of the nodes to comply with
+			// the PotentialNodes length limit. Randomization is
+			// done for us by Go which iterates over map entries
+			// randomly.
+			nodeNames := map[string]struct{}{}
+			for _, node := range nodes {
+				nodeNames[node.Name] = struct{}{}
+			}
+			for nodeName := range nodeNames {
+				if len(podScheduling.Spec.PotentialNodes) >= resourcev1alpha1.PodSchedulingNodeListMaxSize {
+					break
+				}
+				podScheduling.Spec.PotentialNodes = append(podScheduling.Spec.PotentialNodes, nodeName)
+			}
+		}
+		sort.Strings(podScheduling.Spec.PotentialNodes)
```

Comment:
do they need to be sorted?

Reply:
No, this is just to make the content more predictable and readable.

Reply:
Also added to https://github.com/kubernetes/kubernetes/issues/113701.